src/Compilers/Core/MSBuildTask/BuildClient.cs

```diff
@@ -52,35 +52,43 @@ private static string GetExpectedServerExeDir()
             RequestLanguage language,
             Func<string[], int> fallbackCompiler)
         {
-            var errorMessage = CommandLineParser.CheckClientArgsForErrors(args);
+            args = args.Select(arg => arg.Trim()).ToArray();
 
+            bool hasShared;
+            var errorMessage = CommandLineParser.CheckArgsForClientErrors(args, out hasShared);
             if (errorMessage != null)
             {
                 Console.Out.WriteLine(errorMessage);
                 return CommonCompiler.Failed;
             }
 
-            var responseTask = TryRunServerCompilation(
-                language,
-                Environment.CurrentDirectory,
-                args,
-                default(CancellationToken),
-                libEnvVariable: Environment.GetEnvironmentVariable("LIB"));
-
-            responseTask.Wait();
-
-            int exitCode;
-            var response = responseTask.Result;
-            if (response != null)
-            {
-                exitCode = HandleResponse(response);
-            }
-            else
+            IEnumerable<string> validatedArgs = args;
+            if (hasShared)
             {
-                exitCode = fallbackCompiler(args.Where(arg => !arg.StartsWith("/keepalive", StringComparison.Ordinal)).ToArray());
+                validatedArgs = validatedArgs
+                    .Where(arg => !string.Equals(arg,
+                                                 "/shared",
+                                                 StringComparison.OrdinalIgnoreCase));
+                var responseTask = TryRunServerCompilation(
+                    language,
+                    Environment.CurrentDirectory,
+                    validatedArgs.ToArray(),
+                    default(CancellationToken),
+                    libEnvVariable: Environment.GetEnvironmentVariable("LIB"));
+
+                responseTask.Wait();
+
+                var response = responseTask.Result;
```

Comment:
`Wait()` is redundant. 

[Task<TResult>.Result Property](https://msdn.microsoft.com/en-us/library/dd321468%28v=vs.110%29.aspx)
> Accessing the property's get accessor blocks the calling thread until the asynchronous operation is complete; it is equivalent to calling the Wait method.